### PR TITLE
memory exchanging foundation: queue of call

### LINKED ISSUE #52: baseline benchmark
Test
- [x] instance invokes runtime
    - #54 
    - #58 
- [x] instance invokes instance
- [x] runtime invokes instance

### Tool

- rust nightly benchmark
- reference: https://askeing.github.io/rust-book/benchmark-tests.html

### LINKED ISSUE #71: New memory exchanging foundation
> Thanks @L-jasmine for his idea

## Basic idea

Using runtime(wasmedge) memory operation to replace all `read`, `allocate`, after all memory are copied, the real function only send & get addresses.

### instance invokes instance

When invoke `foo a b`, to send `a`, caller
1. get the address & data length of `a`
2. using `wit_write_b a_addr a_len` (this is provided by runtime)
runtime is going to expand the function
3. runtime allocate `a_len` on callee instance, and have an address
4. runtime copy data from `a_addr` to the address of callee

Done

The nice part in this case is, reverse case to put result is the same process.

### instance invokes runtime

This case is `wit_write a_addr a_len` then done (runtime own instance memory).

Reversely, runtime allocate on the caller instance to put result.

### runtime invokes instance

In this case, runtime can put argument into instance, and get result from read memory freely.

## Conclusion

This is a good direction to reduce overhead that many `read`/`write` invocation, the result should be better performance & more simpler codegen. However, to ensure each component has a dual process to be workable with codegen, above analytics must be elaborated to make it work.

### LINKED ISSUE #76: stop rely on wasm-abi
The idea is #71 will reduce more interface, we might have a chance to stop rely on [Rust feature: wasm-abi](https://doc.rust-lang.org/beta/unstable-book/language-features/wasm-abi.html), to work with stable Rust.

NOTE:
1. `#[link(wasm_import_module = "mod")]` is not related to wasm-abi

### PR DESCRIPTION
**Pull request by dannypsnl**
rust bindings runtime: add three host functions
1. require_queue (each call will need to require a new queue)
2. write (write arguments/returns into queue)
3. read (read from queue)

Each caller do
1. requires a queue, get an ID
2. write all parameters to queue by ID
3. invokes convention function, send ID to it
4. read all returns back from queue by ID

Each callee (implementor) do
1. read all arguments back from queue by ID
2. compute via implement function
3. write returns back to queue by ID

resolve #71
resolve #76 
resolve #52

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

File: src/Wit/Gen/Import.hs [hunk 1]
```diff
@@ -23,21 +23,23 @@ toVmWrapper importName = \case
...
-            ( [ pretty "let cfg = CallingConfig::new" <+> tupled [pretty "vm", dquotes $ pretty importName] <+> pretty ";",
-                pretty "let mut args = vec![];"
-              ]
-                ++ map
-                  (\(p, _) -> pretty "let mut a = cfg.put_to_remote" <+> parens (pretty "&" <+> pretty p) <+> pretty ";" <+> pretty "args.append(&mut a);")
-                  param_list
-                ++ [ pretty "let r = cfg.run" <+> tupled [dquotes $ pretty $ externalConvention name, pretty "args"] <+> pretty ";",
-                     pretty "let result_len = r[1].to_i32() as usize;",
-                     pretty "let mut s = String::with_capacity(result_len);",
-                     pretty "cfg.read_from_remote(&mut s, r[0], result_len)"
```
Comment (dannypsnl) [inline src/Wit/Gen/Import.hs:36]: A bug here is `cfg.run` didn't get ported

File: bench/runtime/src/lib.rs [hunk 2]
```diff
@@ -76,7 +84,33 @@ mod tests {
...
+        b.iter(|| {
+            base2(
+                &vm,
+                c2 {
+                    name: "test".to_string(),
+                    age: 1,
+                },
+            );
         });
     }
```
Comment (dannypsnl) [inline bench/runtime/src/lib.rs:116]: Just a note:
1. runtime invokes instance & instance invokes runtime have basically same level performance
2. instance invokes instance will take more more time, around 2x times compares with above

**Review by dm4 — APPROVED** [review]
👍 👀 💯

**Review by dm4 — APPROVED** [review]

**Pull request merged**